File: sorting/quick/quick_sort.rb
def swap(arr, i_index, j_index)
  temp = arr[i_index]
  arr[i_index] = arr[j_index]
  arr[j_index] = temp

  arr
end

# selects an item (pivot) and split the rest interms of
# <= pivot >=
def partition(arr, a_start, a_end)
  pivot = arr[a_end]
  p_index = a_start

  # iterates through moving items <= pivot, to the beginning
  # p_index is where we'll place our pivot
  for i in a_start..(a_end - 1) do
    if arr[i] <= pivot
      swap(arr, i, p_index)
      p_index += 1
    end
  end

  # move pivot to p_index position
  swap(arr, a_end, p_index)

  p_index
end

def quicksort(arr, a_start, a_end)
  return if a_start >= a_end

  pivot = partition(arr, a_start, a_end)

  # pivot - 1 and pivot + 1 is because pivot is already sorted
  quicksort(arr, a_start, pivot - 1)
  quicksort(arr, pivot + 1, a_end)
end

a = [1, 8, 3, 4, 7, -2, 5]

quicksort(a, 0, a.length - 1)

p a

File: sorting/quick/Quicksort.java
package sorting.quick;

import java.util.Arrays;

public class Quicksort {
  private static void swap(int[] arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  private static int partition(int[] arr, int start, int end) {
    int pValue = arr[end];
    int pIndex = start;

    for (int i = start; i <= end - 1; i++) {
      if (arr[i] <= pValue) {
        swap(arr, i, pIndex++);
      }
    }

    swap(arr, end, pIndex);

    return pIndex;
  }

  private static void quicksort(int[] arr, int start, int end) {
    if (start >= end) {
      return;
    }

    int pivot = partition(arr, start, end);

    quicksort(arr, start, pivot - 1);
    quicksort(arr, pivot + 1, end);
  }

  public static void main(String[] args) {
    int[] arr = { 10, 9, 6, 5, 8, -1, -6, 2, -44 };
    quicksort(arr, 0, arr.length - 1);

    System.out.println(Arrays.toString(arr));
  }
}